client/components/tinymce/plugins/mentions/suggestion-list.jsx

```diff
@@ -0,0 +1,90 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+/**
+ * Internal dependencies
+ */
+import PopoverMenu from 'components/popover/menu';
+import PopoverMenuItem from 'components/popover/menu-item';
+import Suggestion from './suggestion';
+
+const SuggestionList = React.createClass( {
+	displayName: 'SuggestionList',
+
+	propTypes: {
+		query: React.PropTypes.string,
+		suggestions: React.PropTypes.array,
+		isVisible: React.PropTypes.bool,
+		onClick: React.PropTypes.func,
+		onClose: React.PropTypes.func
+	},
+
+	getDefaultProps() {
+		return {
+			query: '',
+			suggestions: [],
+			isVisible: false
+		};
+	},
+
+	getMatchingSuggestions() {
+		const { suggestions, query } = this.props;
+		const matcher = new RegExp( "^" + query + "| " + query, 'ig' ); // Start of string or preceded by a space.
+		let matchingSuggestions = suggestions;
+
+		if ( query.length > 0 ) {
+			matchingSuggestions = [];
+
+			for ( let i = 0, len = suggestions.length; i < len; i++ ) {
```

Comment:
I agree with @aduth but would like to extend the notion because our array operations are safe on empty lists. that means we can skip the check for `query.length` because it's built-in

```js
const suggestionName = suggestion => 
	( suggestion.name || `${ suggestion.user_login } ${ suggestion.display_name }` )
		.toLowerCase();

getMatchingSuggestions() {
	const { suggestions, query } = this.props;
	const matcher = new RegExp( `^${ query }|\s${ query }`, 'ig' );

	return suggestions
		.filter( s => matcher.test( suggestionName( s ) ) // only those whose names match
		.slice( 0, 10 ); // and only return up to the first ten results
}
```